### PR TITLE
feat(star-adventurer-gti): Phase 2 BDD scaffold + skywatcher-motor-protocol crate

### PR DESCRIPTION
**Pull request by ivonnyssen**
## Summary

Phase 2 of the Sky-Watcher Star Adventurer GTi Telescope driver, following the design-first / test-first workflow. Lands the test-first scaffold that Phase 3 implementation will turn green.

- New crate `skywatcher-motor-protocol`: pure codec for the Sky-Watcher motor-controller wire protocol (USB + UDP/11880). Types defined; encode/decode bodies are `unimplemented!()` placeholders for Phase 3. Reusable for any Sky-Watcher mount that speaks this protocol (AZ-GTi, EQ6-R, etc.).
- New service `star-adventurer-gti`: ASCOM Alpaca Telescope skeleton matching the qhy-focuser architectural sibling. `Config` (with tagged USB/UDP transport enum), `Transport` trait + serial/udp/mock stubs, `TransportManager`, `coordinates.rs`, `MountDevice` implementing `ascom-alpaca`'s `Telescope` trait — capability flags constant per the design doc, abstract methods stub at `NOT_IMPLEMENTED` or sensible defaults.
- 9 BDD feature files, all `@wip`, 54 scenarios derived from the design doc. Contract constants live in Gherkin tables per `testing.md` §2.5. `bdd.rs` uses `bdd_infra::bdd_main!` + `filter_run_and_exit` to skip `@wip` so the green-suite invariant holds. Step definitions compile against the stub types; bodies use `todo!()` for the parts Phase 3 will implement.
- Workspace integration: both packages wired into `Cargo.toml` + `workspace.dependencies`, `docs/workspace.md` updated, design doc gets a phase-status table.

Phase 3 will: implement the codec, transports, transport manager, coordinate math, and mount-device methods; remove the `@wip` tags as scenarios go green.

## Test plan

- [x] `cargo rail run --profile commit -q` → 1388 tests pass, 0 failures
- [x] `cargo clippy --all --all-targets --all-features -- -D warnings` clean (pre-commit hook)
- [x] `cargo build --all-features --all-targets --locked` clean
- [x] `cargo fmt` applied
- [ ] Phase 3 will turn the 54 `@wip` scenarios green and remove the tags

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: services/star-adventurer-gti/tests/bdd/steps/tracking_steps.rs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+#[when("I try to enable tracking")]
+async fn try_enable_tracking(world: &mut StarAdventurerWorld) {
+    match world.mount().set_tracking(true).await {
+        Ok(()) => world.last_error = None,
+        Err(e) => {
+            world.last_error_code = Some(e.code.raw());
+            world.last_error = Some(e.message.to_string());
+        }
+    }
```
[the remark this answers is not in the document]
Comment (ivonnyssen) [reply services/star-adventurer-gti/tests/bdd/steps/tracking_steps.rs:23]: Agreed — fixed in ac7107d. Added `World::clear_error` and `World::record_error` helpers; rewrote all 13 `try_*` step bodies to call them, so the success branch now clears both `last_error` and `last_error_code`.

File: services/star-adventurer-gti/tests/bdd/steps/sync_steps.rs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+#[when(expr = "I try to sync to RA {float} hours and Dec {float} degrees")]
+async fn try_sync_to(world: &mut StarAdventurerWorld, ra: f64, dec: f64) {
+    match world.mount().sync_to_coordinates(ra, dec).await {
+        Ok(()) => world.last_error = None,
+        Err(e) => {
+            world.last_error_code = Some(e.code.raw());
+            world.last_error = Some(e.message.to_string());
+        }
+    }
```
[the remark this answers is not in the document]
Comment (ivonnyssen) [reply services/star-adventurer-gti/tests/bdd/steps/sync_steps.rs:18]: Agreed — fixed in ac7107d. Added `World::clear_error` and `World::record_error` helpers; rewrote all 13 `try_*` step bodies to call them, so the success branch now clears both `last_error` and `last_error_code`.

File: services/star-adventurer-gti/tests/bdd/steps/slew_steps.rs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+#[when(expr = "I try to slew asynchronously to RA {float} hours and Dec {float} degrees")]
+async fn try_slew_async_to(world: &mut StarAdventurerWorld, ra: f64, dec: f64) {
+    match world.mount().slew_to_coordinates_async(ra, dec).await {
+        Ok(()) => world.last_error = None,
+        Err(e) => {
+            world.last_error_code = Some(e.code.raw());
+            world.last_error = Some(e.message.to_string());
+        }
+    }
```
[the remark this answers is not in the document]
Comment (ivonnyssen) [reply services/star-adventurer-gti/tests/bdd/steps/slew_steps.rs:28]: Agreed — fixed in ac7107d. Added `World::clear_error` and `World::record_error` helpers; rewrote all 13 `try_*` step bodies to call them, so the success branch now clears both `last_error` and `last_error_code`.

File: services/star-adventurer-gti/tests/bdd/steps/park_steps.rs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+#[when("I try to park the mount")]
+async fn try_park_mount(world: &mut StarAdventurerWorld) {
+    match world.mount().park().await {
+        Ok(()) => world.last_error = None,
+        Err(e) => {
+            world.last_error_code = Some(e.code.raw());
+            world.last_error = Some(e.message.to_string());
+        }
+    }
```
[the remark this answers is not in the document]
Comment (ivonnyssen) [reply services/star-adventurer-gti/tests/bdd/steps/park_steps.rs:18]: Agreed — fixed in ac7107d. Added `World::clear_error` and `World::record_error` helpers; rewrote all 13 `try_*` step bodies to call them, so the success branch now clears both `last_error` and `last_error_code`.

File: services/star-adventurer-gti/tests/bdd/steps/coordinate_steps.rs [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+#[when("I try to read RightAscension")]
+async fn try_read_ra(world: &mut StarAdventurerWorld) {
+    match world.mount().right_ascension().await {
+        Ok(_) => world.last_error = None,
+        Err(e) => {
+            world.last_error_code = Some(e.code.raw());
+            world.last_error = Some(e.message.to_string());
+        }
+    }
```
[the remark this answers is not in the document]
Comment (ivonnyssen) [reply services/star-adventurer-gti/tests/bdd/steps/coordinate_steps.rs:48]: Agreed — fixed in ac7107d. Added `World::clear_error` and `World::record_error` helpers; rewrote all 13 `try_*` step bodies to call them, so the success branch now clears both `last_error` and `last_error_code`.

File: services/star-adventurer-gti/tests/bdd/steps/side_of_pier_steps.rs [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+#[when(expr = "I try to read DestinationSideOfPier for RA {float} hours and Dec {float} degrees")]
+async fn try_read_destination_side(world: &mut StarAdventurerWorld, ra: f64, dec: f64) {
+    match world.mount().destination_side_of_pier(ra, dec).await {
+        Ok(_) => world.last_error = None,
+        Err(e) => {
+            world.last_error_code = Some(e.code.raw());
+            world.last_error = Some(e.message.to_string());
+        }
+    }
```
[the remark this answers is not in the document]
Comment (ivonnyssen) [reply services/star-adventurer-gti/tests/bdd/steps/side_of_pier_steps.rs:27]: Agreed — fixed in ac7107d. Added `World::clear_error` and `World::record_error` helpers; rewrote all 13 `try_*` step bodies to call them, so the success branch now clears both `last_error` and `last_error_code`.

File: services/star-adventurer-gti/tests/bdd/steps/abort_steps.rs [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+#[when("I try to abort the slew")]
+async fn try_abort_slew(world: &mut StarAdventurerWorld) {
+    match world.mount().abort_slew().await {
+        Ok(()) => world.last_error = None,
+        Err(e) => {
+            world.last_error_code = Some(e.code.raw());
+            world.last_error = Some(e.message.to_string());
+        }
+    }
```
[the remark this answers is not in the document]
Comment (ivonnyssen) [reply services/star-adventurer-gti/tests/bdd/steps/abort_steps.rs:18]: Agreed — fixed in ac7107d. Added `World::clear_error` and `World::record_error` helpers; rewrote all 13 `try_*` step bodies to call them, so the success branch now clears both `last_error` and `last_error_code`.

File: services/star-adventurer-gti/src/transport/mock.rs [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    /// `0x375F00` (3,628,800) is the GTi default; tests can override.
+    pub cpr_ra: u32,
+    pub cpr_dec: u32,
+    /// `0xF42400` (≈ 16 MHz) is the GTi default.
+    pub tmr_freq: u32,
+    pub high_speed_ratio_ra: u32,
+    pub high_speed_ratio_dec: u32,
+    /// `0x03300C` matches the GTi probe table in the design doc.
+    pub motor_board_version: u32,
+}
```
[the remark this answers is not in the document]
Comment (ivonnyssen) [reply services/star-adventurer-gti/src/transport/mock.rs:72]: Agreed — fixed in ac7107d. Replaced `derive(Default)` with a manual impl that seeds CPR (`0x375F00` on both axes), TMR_Freq (`0xF42400`), motor-board version (`0x03300C`), and a reasonable high-speed-ratio default (`32`) per the GTi probe table. Tests can still override individual fields.

File: crates/skywatcher-motor-protocol/src/error.rs [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+}
+
+impl MountErrorCode {
+    /// Decode a single error nibble (`0`..=`F`) into a [`MountErrorCode`].
+    ///
+    /// `code` is the parsed numeric value of the error nibble, not the ASCII
+    /// byte; callers that have raw hex bytes should run them through
+    /// [`crate::codec::decode_u8`] first.
+    pub fn from_nibble(code: u8) -> Self {
+        match code {
```
[the remark this answers is not in the document]
Comment (ivonnyssen) [reply crates/skywatcher-motor-protocol/src/error.rs]: Agreed — fixed in ac7107d. Renamed `MountErrorCode::from_nibble` → `from_byte` and tightened the docstrings on `ProtocolError::MountError`, `MountErrorCode::Unknown`, and `from_byte` itself to consistently say "byte" (two hex digits in the `!XX\r` wire frame). The named variants are still single-nibble values per the spec; everything else maps to `Unknown(u8)`.

**Comment by ivonnyssen** [comment]
Coverage delta is expected for this PR — Phase 2 scaffolds the `skywatcher-motor-protocol` crate with deliberately-uncovered `unimplemented!()` stub functions (`encode_u24`, `decode_position`, `Command::encode_into`, `Response::decode`, etc.). Phase 3 will land the codec implementation alongside crate-level unit + property tests (round-trip invariants, codec edge cases) which the design doc calls out in §Testing. The 0% on these specific files is the right number until then; the project as a whole stays at 92.8%.

File: services/star-adventurer-gti/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,177 @@
+#![cfg_attr(coverage_nightly, feature(coverage_attribute))]
+//! Star Adventurer GTi ASCOM Alpaca driver.
+//!
+//! See [`docs/services/star-adventurer-gti.md`](../../docs/services/star-adventurer-gti.md)
```
[the remark this answers is not in the document]
Comment (ivonnyssen) [reply services/star-adventurer-gti/src/lib.rs]: Agreed — fixed in 9a8a1c1. Bumped `../../docs/...` to `../../../docs/...` to match `src/config.rs` and the rest of the workspace.

File: services/star-adventurer-gti/src/transport/mock.rs [hunk 11]
```diff
@@ -0,0 +1,157 @@
+//! Feature-gated in-memory mock transport.
+//!
+//! Simulates the motor controller as a small state machine: accepts
+//! `:cmd<axis><payload>\r` frames, maintains per-axis state (position,
+//! motion mode, running flag, initialised flag, tracking), and emits
+//! well-formed `=...\r` / `!XX\r` replies. Used by:
+//!
+//! * BDD tests (via [`crate::ServerBuilder::with_transport`])
+//! * `tests/test_lib.rs` server-startup tests
+//! * the `conformu` integration target
```
[the remark this answers is not in the document]
Comment (ivonnyssen) [reply services/star-adventurer-gti/src/transport/mock.rs]: Agreed — fixed in 9a8a1c1. Rewrote the doc to name only the real Phase 2 consumer (`tests/bdd.rs`) and explicitly mark `tests/test_lib.rs` and the `conformu` integration target as Phase 3 follow-ups, not current code.

File: services/star-adventurer-gti/src/transport_manager.rs [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+        unimplemented!(
+            "Phase 3: decrement count, on zero stop poll task, abort motion, close transport"
+        )
+    }
+
+    /// `true` when the transport is currently open (handshake completed and
+    /// not yet torn down).
+    pub fn is_available(&self) -> bool {
+        self.connection_count.load(Ordering::SeqCst) > 0
+    }
```
[the remark this answers is not in the document]
Comment (ivonnyssen) [reply services/star-adventurer-gti/src/transport_manager.rs]: Agreed — fixed in 9a8a1c1. Added an `available: AtomicBool` field (mirroring `qhy-focuser::SerialManager::serial_available`) that Phase 3 will set only after the init handshake succeeds. `is_available()` now reads from it, so a connect-in-flight or a handshake-failure rollback can no longer false-positive. Updated the connect/disconnect doc comments to call out where the bit gets flipped.

File: services/star-adventurer-gti/src/error.rs [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+impl StarAdvError {
+    /// Map a driver error to the closest ASCOM error code.
+    pub fn to_ascom_error(self) -> ASCOMError {
+        match self {
+            Self::NotConnected => ASCOMError::new(ASCOMErrorCode::NOT_CONNECTED, self.to_string()),
+            Self::InvalidValue(_) => {
+                ASCOMError::new(ASCOMErrorCode::INVALID_VALUE, self.to_string())
+            }
+            Self::Parked => ASCOMError::new(ASCOMErrorCode::INVALID_WHILE_PARKED, self.to_string()),
+            _ => ASCOMError::invalid_operation(self.to_string()),
```
[the remark this answers is not in the document]
Comment (ivonnyssen) [reply services/star-adventurer-gti/src/error.rs]: Refactored in 085eaa3 — pulled `self.to_string()` out of the match and reused the resulting `msg` in each arm. The original compiled and passed clippy (Rust is happy with `self.to_string()` in arms because the unit / `_`-bound variants don't move any fields out of `self`), but the cleaner shape avoids re-rendering the Display impl four times and matches what you'd write idiomatically.

File: services/star-adventurer-gti/tests/bdd/steps/connection_steps.rs [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+async fn param_cache_cpr_dec(world: &mut StarAdventurerWorld, expected: u32) {
+    todo!("Phase 3: read TransportManager.parameters().cpr_dec and assert eq")
+}
+
+#[then(expr = "the parameter cache should report timer frequency {int}")]
+async fn param_cache_tmr_freq(world: &mut StarAdventurerWorld, expected: u32) {
+    todo!("Phase 3: read TransportManager.parameters().tmr_freq and assert eq")
+}
+
+#[then(expr = "the mount should have received command {word}")]
```
[the remark this answers is not in the document]
Comment (ivonnyssen) [reply services/star-adventurer-gti/tests/bdd/steps/connection_steps.rs:99]: Verified against the source — cucumber-expressions 0.5.0 (the version the workspace pins via `cucumber = "0.22.1"`) defines `{word}` as the regex `[^\\s]+` (see `cucumber-expressions-0.5.0/src/expand/mod.rs:434`). `:K1`, `:L1`, `:S1` etc. are all non-whitespace, so they match `{word}` cleanly. Leaving the step as-is.

File: services/star-adventurer-gti/Cargo.toml [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+tracing = { workspace = true }
+tracing-subscriber = { workspace = true }
+thiserror = { workspace = true }
+rp-tls = { workspace = true }
+rp-auth = { workspace = true }
+axum = { workspace = true }
+
+[package.metadata.conformu]
+command = "cargo test -p star-adventurer-gti --features conformu --test conformu_integration -- --ignored --nocapture"
+
```
[the remark this answers is not in the document]
Comment (ivonnyssen) [reply services/star-adventurer-gti/Cargo.toml]: Agreed — fixed in 085eaa3. Removed the `[package.metadata.conformu]` block until `tests/conformu_integration.rs` lands in Phase 3 (verified that every other opted-in service — qhy-focuser, ppba-driver, filemonitor, sky-survey-camera — ships the test file together with the metadata).

File: crates/skywatcher-motor-protocol/src/codec.rs [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+/// Verify that `frame` matches the strict UDP framing rules: starts with `:`,
+/// ends with a single `\r`, and contains no junk before/after.
+///
+/// Used by the UDP transport on receive. Serial transports buffer up to the
+/// first `\r` and pass exactly the resulting slice in here.
+#[cfg_attr(coverage_nightly, coverage(off))]
+pub fn validate_frame(_frame: &[u8]) -> Result<()> {
+    unimplemented!("Phase 3: enforce one well-formed `:...\\r` frame, nothing trailing")
```
[the remark this answers is not in the document]
Comment (ivonnyssen) [reply crates/skywatcher-motor-protocol/src/codec.rs]: Agreed — fixed in f1f19ff. Split into two functions: `validate_command_frame` (outbound `:cmd<axis><payload?>\\r`) and `validate_response_frame` (inbound `=...\\r` / `!XX\\r`). Both are still Phase 2 `unimplemented!()` stubs with `coverage(off)`, but the API surface now matches the actual wire shapes the codec needs to police on each direction.

File: services/star-adventurer-gti/src/lib.rs [hunk 17]
```diff
@@ -0,0 +1,177 @@
...
+        server.discovery_port = self.config.server.discovery_port;
+
+        // Phase 3 wires real transports here. Until then the only path
+        // that boots a server is one that injects a transport explicitly
+        // (see BDD tests).
+        let transport = self.transport.ok_or_else(|| {
+            Box::<dyn std::error::Error>::from(
+                "no transport injected — Phase 3 will pick serial/UDP from config.transport",
+            )
+        })?;
```
[the remark this answers is not in the document]
Comment (ivonnyssen) [reply services/star-adventurer-gti/src/lib.rs]: Agreed — fixed in f1f19ff together with the TransportManager comment. Introduced a `TransportFactory` trait (mirroring `qhy-focuser::SerialPortFactory`); when `ServerBuilder` is built without an injected factory it now derives one from `config.transport` (`Usb` → `SerialTransportFactory`, `Udp` → `UdpTransportFactory`). Production binary now boots cleanly without `--features mock` — ASCOM `Connected = true` still returns NOT_IMPLEMENTED until Phase 3 fills the factory bodies in, but the HTTP server binds and metadata/capability endpoints respond.

File: services/star-adventurer-gti/src/transport_manager.rs [hunk 18]
```diff
@@ -0,0 +1,172 @@
...
+}
+
+/// Shared, ref-counted transport handle.
+pub struct TransportManager {
+    _config: Config,
+    transport: Arc<dyn Transport>,
+    /// Number of clients that currently believe themselves connected.
+    /// Incremented on the way into [`connect`] and decremented on the way
+    /// out of [`disconnect`]; the actual transport open/close is gated on
+    /// the count crossing 0.
```
[the remark this answers is not in the document]
Comment (ivonnyssen) [reply services/star-adventurer-gti/src/transport_manager.rs:66]: Agreed — fixed in f1f19ff (same commit as the related ServerBuilder comment). `TransportManager` now stores `factory: Arc<dyn TransportFactory>` plus `transport: Mutex<Option<Arc<dyn Transport>>>`. The 0→1 connect transition calls `factory.open(&config)` (Phase 3 then runs the handshake); the 1→0 disconnect transition clears the `Mutex<Option<...>>` slot, dropping the `Arc` and triggering `Transport::close`. Matches the qhy-focuser pattern you pointed at.

File: services/star-adventurer-gti/src/transport/mock.rs [hunk 19]
```diff
@@ -0,0 +1,172 @@
+//! Feature-gated in-memory mock transport.
+//!
+//! Simulates the motor controller as a small state machine: accepts
+//! `:cmd<axis><payload>\r` frames, maintains per-axis state (position,
+//! motion mode, running flag, initialised flag, tracking), and emits
+//! well-formed `=...\r` / `!XX\r` replies. Phase 2 wires it through
+//! [`crate::ServerBuilder::with_transport`] for the BDD `tests/bdd.rs`
```
[the remark this answers is not in the document]
Comment (ivonnyssen) [reply services/star-adventurer-gti/src/transport/mock.rs]: Agreed — fixed in 16f38c4. Updated the rustdoc link to `with_transport_factory` (left over from the round 4 rename).
